app/src/main/java/com/example/locavore/LoginActivity.java

```diff
@@ -2,13 +2,89 @@
 
 import androidx.appcompat.app.AppCompatActivity;
 
+import android.content.Intent;
 import android.os.Bundle;
+import android.util.Log;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.Toast;
+
+import com.parse.ParseUser;
 
 public class LoginActivity extends AppCompatActivity {
 
+    public static final String TAG = "LoginActivity";
+    private EditText etEmailAddress;
+    private EditText etPassword;
+    private Button btnLogin;
+    private Button btnFarmerSignup;
+    private Button btnLocavoreSignup;
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_login);
+
+        if (ParseUser.getCurrentUser() != null)
+        {
+            goToMainActivity();
+        }
+
+        etEmailAddress = findViewById(R.id.etUsername);
+        etPassword = findViewById(R.id.etPassword);
+
+        btnFarmerSignup = findViewById(R.id.btnFarmerSignup);
+        btnFarmerSignup.setOnClickListener(v -> goToFarmerSignupActivity());
+
+        btnLocavoreSignup = findViewById(R.id.btnLocavoreSignup);
+        btnLocavoreSignup.setOnClickListener(v -> goToLocavoreSignupActivity());
+
+        btnLogin = findViewById(R.id.btnLogin);
+        btnLogin.setOnClickListener(v -> loginUser(etEmailAddress.getText().toString(), etPassword.getText().toString()));
+    }
+    private void goToFarmerSignupActivity() {
+        Intent i = new Intent(this, FarmerSignupActivity.class);
+        startActivity(i);
+    }
+
+    private void goToLocavoreSignupActivity() {
+        Intent i = new Intent(this, LocavoreSignupActivity.class);
+        startActivity(i);
+    }
+
+    private void goToMainActivity() {
+        Intent i = new Intent(this, MainActivity.class);
+        startActivity(i);
+        finish();
+    }
+
+    private void loginUser(String username, String password) {
```

Comment:
There seems to be decent amount of overlap in the login/signup logic. Especially in the error handling. You may want to create a utility class that can be used across activities/fragments so you don't duplicate loigc.

Reply:
That makes a lot of sense. Thank you!